client/src/main/java/org/apache/celeborn/client/ShuffleClientImpl.java

```diff
@@ -617,22 +630,26 @@ private ConcurrentHashMap<Integer, PartitionLocation> registerShuffleInternal(
           }
           return result;
         } else if (StatusCode.SLOT_NOT_AVAILABLE.equals(respStatus)) {
+          lastException = new CelebornIOException("Slot not available for shuffle " + shuffleId);
           logger.error(
               "LifecycleManager request slots return {}, retry again, remain retry times {}.",
               StatusCode.SLOT_NOT_AVAILABLE,
               numRetries - 1);
         } else if (StatusCode.RESERVE_SLOTS_FAILED.equals(respStatus)) {
+          lastException = new CelebornIOException("Reserve slots failed for shuffle " + shuffleId);
           logger.error(
               "LifecycleManager request slots return {}, retry again, remain retry times {}.",
               StatusCode.RESERVE_SLOTS_FAILED,
               numRetries - 1);
         } else {
+          lastException = new CelebornIOException("Request slots failed for shuffle " + shuffleId);
           logger.error(
               "LifecycleManager request slots return {}, retry again, remain retry times {}.",
               StatusCode.REQUEST_FAILED,
               numRetries - 1);
         }
       } catch (Exception e) {
+        lastException = e;
```

Comment:
@s0nskar @SteNicholas  I use an exception instead of just an error message because at line 652, we should record an unknown exception here. If we have both an error message and an unknown exception, we won't know which was the last cause. We could use numRetries to mark it, but that would be more complex.